### ISSUE TITLE
[process] Output from terminal process is swallowed

### ISSUE THREAD

**Issue by JanKoehnlein** [theia-ide/theia#3791]
When a TerminalProcess is started, we spawn the command in a node-pty terminal and then connect the output. If the process is fast enough, it can produce output before it's connected, thus this output is swallowed. This is particularly bad when the process exits early with an error message.

To reproduce this, create a task that runs a shell script that just echos a string, e.g.
```tasks.json
{
    "tasks": [
        {
            "label": "Test task",
            "type": "shell",
            "cwd": "${workspaceFolder}",
            "command": "./test.sh",
            "args": [
            ]
        }
    ]
}
```

```test.sh
#!/bin/bash
#sleep 2
echo You won\'t see me
```

With the sleep command commented out, you will not see any output in Theia's terminal.

**Comment by svenefftinge** [comment]
Offline discussion results: It should be fixed in the 'tasks' implementation. It should use ShellProcess instead of TerminalProcess and not pass a shell. But instead send the command through `write` on the returned object.

**Comment by simark** [comment]
The problem we identified in #2961 was that for a short lived process, we registered it (on creation) and unregister it (on exit) very quickly from the process manager.  By the time the terminal finally opens in the frontend, the process has been unregistered, and the terminal just runs an interactive shell as a fallback (I guess that's what you see?).

The solution we had in mind was more to either

1. somehow wait for the terminal to be attached before running the process
2. keep the output buffer around until we consume it

I would tend towards option 2.

**Comment by JanKoehnlein** [comment]
Sounds like a different problem: What I am trying to fix is that the output buffer does not get populated at all because the command has already finished before the output buffer is populated. The solution to this is 
1) start the shell
2) attach the output buffer 
3) run command

instead of 

1) run the command in a shell
2) attach the output buffer

**Comment by JanKoehnlein** [comment]
Yes, that's what I see. I faintly remember I also had longer running jobs whose first lines of output were swallowed, but I am not entirely sure anymore. But I get the same error message on the console as you. So if you have a good fix for that, go ahead. 

If you look at the code of the [TerminalProcess](https://github.com/theia-ide/theia/blob/ec57ef562cdd4bcae0daacd3cff59376b47cf306/packages/process/src/node/terminal-process.ts#L49), the method `spawn` is already called with the command (not the shell). After that the output buffer is attached. I thought, this race between the process and the output attachment causes the missing lines. The proposed fix *always* opens an interactive shell, but doesn't loose any output (tested on Windows and Mac). Maybe there is more to it.

Furthermore, the documentation also suggests that on windows I should specify `cmd.exe` as the command, while on other OSs I did not have to specify any shell. The proposed change also fixes this asymmetry and brings us closer to the VSCode task format.

**Comment by simark** [comment]
> So if you have a good fix for that, go ahead.

I haven't looked into it yet, just thought about the possible paths mentioned above.

> If you look at the code of the [TerminalProcess](https://github.com/theia-ide/theia/blob/ec57ef562cdd4bcae0daacd3cff59376b47cf306/packages/process/src/node/terminal-process.ts#L49), the method `spawn` is already called with the command (not the shell). After that the output buffer is attached. I think, a race between the process and the output attachment can cause the loss of output lines.

The way this is designed, it's not supposed to happen.  It works the same way with node's `child_process.spawn`, you register a data callback after having created the process.  Between the time you get the return value in `this.terminal` and the time you register the callback, the execution does not go back to the event loop, so there should be no way that the data output by the process would be consumed.  If there is indeed a window where the data gets consumed before we have the change to register the callback, I'd say it's a bug with node-pty.

> Furthermore, the documentation also suggests that on windows I should specify cmd.exe as the command, while on other OSs I did not have to specify any shell. The proposed change also fixes this asymmetry and brings us closer to the VSCode task format.

Which documentation?

I agree we are quite far from how VSCode handles and executes tasks (e.g. #2960) and it's also my goal to bring us closer.

**Comment by simark** [comment]
Ideally, our process API should be flexible enough to support these two axis of configuration independently:

- What to use as standard input/output: a tty, pipes or nothing
- Should the command be run through a shell, or directly as a process

**Comment by JanKoehnlein** [comment]
The documentation: README.md in the task extension's base dir.

When the command is run as a new process, does it really matter what the JS event loop does? I wouldn't claim the missing lines are consumed. I assume they are just missed. Of course I may be wrong. Sounds like you have dug deeper into the node-pty code. I won't even rule out it has bugs. At least the example in the README of node-pty does it the way I designed my fix.

I am open to any kind of solution to this. I applied my fix for a customer project and it worked well enough for them. It was much more important to see the full output (especially on errors) than to avoid an interactive terminal.

**Comment by sgraband** [comment]
Ring buffer prevents output loss.
